smallrye-reactive-messaging-mqtt/src/main/java/io/smallrye/reactive/messaging/mqtt/MqttSink.java

```diff
@@ -13,94 +13,83 @@
-    private final int port;
-    private final String server;
     private final String topic;
     private final int qos;
 
     private final SubscriberBuilder<? extends Message<?>, Void> sink;
-    private final AtomicBoolean connected = new AtomicBoolean();
+    private final AtomicBoolean ready = new AtomicBoolean();
 
     public MqttSink(Vertx vertx, MqttConnectorOutgoingConfiguration config) {
-        MqttClientOptions options = MqttHelpers.createMqttClientOptions(config);
-        host = config.getHost();
-        int def = options.isSsl() ? 8883 : 1883;
-        port = config.getPort().orElse(def);
-        server = config.getServerName().orElse(null);
+        MqttClientSessionOptions options = MqttHelpers.createMqttClientOptions(config);
         topic = config.getTopic().orElseGet(config::getChannel);
         qos = config.getQos();
 
-        AtomicReference<MqttClient> reference = new AtomicReference<>();
+        AtomicReference<Clients.ClientHolder> reference = new AtomicReference<>();
         sink = ReactiveStreams.<Message<?>> builder()
                 .flatMapCompletionStage(msg -> {
-                    MqttClient client = reference.get();
-                    if (client != null) {
-                        if (client.isConnected()) {
-                            connected.set(true);
-                            return CompletableFuture.completedFuture(msg);
-                        } else {
-                            CompletableFuture<Message<?>> future = new CompletableFuture<>();
-                            vertx.setPeriodic(100, id -> {
-                                if (client.isConnected()) {
-                                    vertx.cancelTimer(id);
-                                    connected.set(true);
-                                    future.complete(msg);
-                                }
-                            });
-                            return future;
-                        }
-                    } else {
-                        return Clients.getConnectedClient(vertx, host, port, server, options)
-                                .map(c -> {
-                                    reference.set(c);
-                                    connected.set(true);
-                                    return msg;
-                                })
-                                .subscribeAsCompletionStage();
+                    Clients.ClientHolder client = reference.get();
+                    if (client == null) {
+                        client = Clients.getHolder(vertx, options);
+                        // FIXME: add session state listener
+                        client.start();
+                        reference.set(client);
                     }
+
+                    return client.sessionState()
+                            .filter(state -> state != SessionState.CONNECTED)
+                            .map(ignore -> msg).toUni().subscribeAsCompletionStage();
```

Comment:
That is just waiting for the first item. Is that what you expect?

Reply:
My understanding of this part (and I may be wrong here) is that this code get a "client" for a single operation of sending a messages. Closing it later on, when the sending is complete.

If that understanding is correct, then yes, the first event switching to CONNECTED is sufficient. As this waits for the first time the connection is established, and then it sends the message.

If my assumption is not correct, and the client is kept open. Then that part should probably to the "send" method.